### PR TITLE
added login

### LINKED ISSUE #47: Login and password reset
Login is required after registeration of a user. Hence, adding a login view.

Password reset feature, where an email is sent on the user's email id and user can reset the password.

### PR DESCRIPTION
**Pull request by DhruviJ17**
### Description

Added link of contributing guidelines

Fixes #47 

### Type of Change:

- Code
- Quality Assurance
- User Interface
- Outreach
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### Checklist:

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] I have written Kotlin Docs whenever is applicable

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Review by devkapilbansal — COMMENTED** [review]
@DhruviJ17 which issue you are working on. Please link that issue too in the pull request and enforce issue closing in commit message. If there is no such issue, consider opening one and wait till the issue is accepted and assigned to you.

**Comment by DhruviJ17** [comment]
I am just trying to add new features which I thought were missing!

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Thanks @DhruviJ17 
Please open an issue for this too. Also, edit the pull request message template and fill relevant information there

**Comment by DhruviJ17** [comment]
@devkapilbansal I have opened an issue for the above so please can you accept it and assign it to me

**Comment by devkapilbansal** [comment]
> @devkapilbansal I have opened an issue for the above so please can you accept it and assign it to me

@DhruviJ17 I can't accept it. Consider asking it on zulip

**Comment by isabelcosta** [comment]
@DhruviJ17 can you please post the link in a comment here, or edit the PR (pull request) description to include that?

**Comment by DhruviJ17** [comment]
@isabelcosta  this is the link for the issue created by me: https://github.com/anitab-org/open-source-programs-backend/issues/47

File: token_auth/views/password_reset.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+from rest_framework.response import Response
+from django.contrib.sites.shortcuts import get_current_site
+from django.contrib.auth.models import User
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+from django.utils.encoding import smart_str, force_str, smart_bytes, DjangoUnicodeDecodeError
+from django.utils.http import urlsafe_base64_decode, urlsafe_base64_encode
+from token_auth.utils import Util
+
+
+class RequestPasswordResetEmailView(generics.GenericAPIView):
```
Comment (isabelcosta) [inline token_auth/views/password_reset.py]: Can you please add a doc string (in code documentation) explaining what this test is testing. E.g.: Given user that is registered, when _______, then ______. Just an example, you don't need to follow this format.

File: token_auth/views/password_reset.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        if User.objects.filter(email=email).exists():
+            user = User.objects.get(email=email)
+            uidb64 = urlsafe_base64_encode(smart_bytes(user.id))
+            token = PasswordResetTokenGenerator().make_token(user)
+            current_site = get_current_site(request=request).domain
+            relativeLink = reverse('password-reset-confirm', kwargs={'uidb64': uidb64, 'token': token})
+            redirect_url = request.data.get('redirect_url', '')
+            absurl = 'http://'+current_site + relativeLink
+            email_body = 'Hello, \n Use link below to reset your password  \n' + \
+                absurl+"?redirect_url="+redirect_url
```
Comment (isabelcosta) [inline token_auth/views/password_reset.py]: Can you use double quotes in a consistent manner, please? By this I mean, if you use double-quote here, then use it everywhere on this file.

File: token_auth/serializers/login.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from django.contrib.auth.models import User
+from rest_framework.exceptions import AuthenticationFailed
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+from rest_framework_simplejwt.tokens import RefreshToken, TokenError
+from django.contrib import auth
+from django.utils.encoding import smart_str, force_str, smart_bytes, DjangoUnicodeDecodeError
+from django.utils.http import urlsafe_base64_decode, urlsafe_base64_encode
+
+
+class LoginSerializer(serializers.ModelSerializer):
```
Comment (isabelcosta) [inline token_auth/serializers/login.py]: Can you please add doc strings here, giving a brief explanation of what this class is for?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Great work @DhruviJ17 ! I've request some minor changes, please let me know when you've addressed them and if you have any questions.

Thank you for your contribution!

**Comment by isabelcosta** [comment]
Can you please add the issue number on the PR description? So its reference in the description @DhruviJ17

File: token_auth/views/password_reset.py (hunk 1 above)
Comment (DhruviJ17) [reply token_auth/views/password_reset.py]: Done

File: token_auth/views/password_reset.py (hunk 2 above)
Comment (DhruviJ17) [reply token_auth/views/password_reset.py]: Done

File: token_auth/serializers/login.py (hunk 3 above)
Comment (DhruviJ17) [reply token_auth/serializers/login.py]: Done

**Comment by codesankalp** [comment]
@DhruviJ17 Please link your pull request with the issue by changing the commit message or pull-request description.
Also, squash your commits :+1: .

**Comment by isabelcosta** [comment]
@DhruviJ17 are you still interested in improving this PR? If so could you please edit the PR description so it describes the changes you added, how you tested this and the issue number that you are fixing.
Can you also edit the issue description to follow the issue templates?

If you are not interested anymore it's completely fine, I can close and make the issue available again :)

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Can you remove the `.DS_Store` files and add a test for this change please?

**Comment by DhruviJ17** [comment]
done👍🏻

**Comment by codesankalp** [comment]
@DhruviJ17  `.DS_Store` file is still present. Can you please remove it.

**Comment by codesankalp** [comment]
@DhruviJ17 instead of `**/.DS_Store` add this in `.gitignore`
```
.DS_Store
.AppleDouble
.LSOverride
```
I have added extra git ignores so that in the future there will be no problem regarding this.

**Comment by codesankalp** [comment]
@DhruviJ17 You have added `.DS_Store`  in `.gitignore` but it is being tracked by git before adding this in `.gitignore` so try to unstage it.
You can use this:
```
git rm --cached -- .DS_Store
git rm --cached -- .token_auth/serializers/.DS_Store
```

Also, I saw that `password_reset.py` doesn't have an extra line at the end. Please add that by following PEP8 guidelines.

File: .gitignore [hunk 4]
```diff
@@ -130,3 +130,8 @@ dmypy.json
 .pyre/
 
 .vscode/
+
+
+.DS_Store
+.AppleDouble
+.LSOverride
```
Comment (devkapilbansal) [inline .gitignore]: There should be a blank line at the end

File: token_auth/serializers/password_reset.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+from rest_framework_jwt.serializers import PasswordField
+
+
+class ResetPasswordEmailRequestSerializer(serializers.Serializer):
+    email = serializers.EmailField(min_length=2)
+    redirect_url = serializers.CharField(max_length=500, required=False)
+    class Meta:
+        fields = ['email']
+
+
```
Comment (devkapilbansal) [inline token_auth/serializers/password_reset.py]: Only one line is required here

File: token_auth/views/login.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+"""
+Function for user login after registration
+"""
+
+class LoginView(generics.GenericAPIView):
+    serializer_class = LoginSerializer
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
Comment (devkapilbansal) [inline token_auth/views/login.py]: Add a blank line here too

Comment (devkapilbansal) [inline token_auth/views/login.py]: I don't think that you need generic view here
```suggestion
class LoginView(APIView):
    def post(self, request):
        serializer = LoginSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        return Response(serializer.data, status=status.HTTP_200_OK)

```

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@DhruviJ17 please use the `commit message guidelines` that we have. The commit message 
 should look like:-
```
type: Short Description

Long Description
```

File: token_auth/views/password_reset.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        email = request.data.get('email', '')
+        if User.objects.filter(email=email).exists():
+            user = User.objects.get(email=email)
+            uidb64 = urlsafe_base64_encode(smart_bytes(user.id))
+            token = PasswordResetTokenGenerator().make_token(user)
+            current_site = get_current_site(request=request).domain
+            relativeLink = reverse('password-reset-confirm', kwargs={'uidb64': uidb64, 'token': token})
+            redirect_url = request.data.get('redirect_url', '')
+            absurl = 'http://'+current_site + relativeLink
+            email_body = 'Hello, \n Use link below to reset your password  \n' + \
```
Comment (isabelcosta) [inline token_auth/views/password_reset.py]: is there a way to put this email template separately?

**Comment by codesankalp** [comment]
@DhruviJ17 Can you fix the merge conflict (solution: rebase your branch) and do the required changes as mentioned by @isabelcosta and @devkapilbansal .

**Comment by DhruviJ17** [comment]
> @DhruviJ17 Can you fix the merge conflict (solution: rebase your branch) and do the required changes as mentioned by @isabelcosta and @devkapilbansal .

When i rebased it showed me already up to date. So i just changed the generic view to api view as mentioned above and added the blank lines.

**Comment by codesankalp** [comment]
@DhruviJ17 how do you rebase because I am seeing your branch is 11 commits behind?

Also, the requested changes are to add a single empty line at the end of the file and removing other empty lines.

**Comment by devkapilbansal** [comment]
> > @DhruviJ17 Can you fix the merge conflict (solution: rebase your branch) and do the required changes as mentioned by @isabelcosta and @devkapilbansal .
> 
> When i rebased it showed me already up to date. So i just changed the generic view to api view as mentioned above and added the blank lines.

@DhruviJ17 before rebasing, update your develop branch first. Running `git fetch upstream develop` will do the needful.

**Comment by isabelcosta** [comment]
@DhruviJ17 could you please edit the description of this PR to describe what you have implemented here? It mentions that you added contribution guidelines link

**Pull request closed without merging**